### DISCUSSION TITLE
Select text from a middle of a link

### DISCUSSION THREAD

**Discussion by PF4Public** [Eloston/ungoogled-chromium#2055] [Ideas]
Does anyone know of a way of preventing the drag-drop feature for links by default?

I mean the internals of a browser. I did grepping through the sources for "drag" "draggable", but could not find any place, where the bahaviour could be modified. Since I'm coming from an older Opera, this inability of selecting text anywhere I want just drives me crazy :(

**Comment by networkException** [comment]
You can disable drag and drop / link clicking by holding <kbd>Alt</kbd>, maybe thats useful to find references in the code

**Reply by PF4Public** [reply]
Thanks! That's already something! Untill recently I was using Alt for resizing and moving windows, that's probably the reason why this useful key slipped away.
I'll continue digging the sources with this hint in mind.

**Comment by tomasz1986** [comment]
Just for the record, there's a script https://github.com/eight04/select-text-inside-a-link-like-opera that emulates the old Opera behaviour. I'm also a former Opera user, and I've been using this one ever since leaving the browser. To be honest, having to press and hold `Alt` every time is kind of annoying.

**Reply by PF4Public** [reply]
Oh! I see I'm not the only one here! Great! Or rather not so great. :)

**Comment by PF4Public** [comment]
**EDIT:** implemented in https://github.com/ungoogled-software/ungoogled-chromium/pull/2080


Found it! Thanks to @networkException's hint. It is in [third_party/blink/renderer/core/editing/selection_controller.cc](https://source.chromium.org/chromium/chromium/src/+/refs/tags/105.0.5195.52:third_party/blink/renderer/core/editing/selection_controller.cc;l=1321-1325;bpv=0;bpt=1)

```c
bool IsSelectionOverLink(const MouseEventWithHitTestResults& event) {
  return (event.Event().GetModifiers() & WebInputEvent::Modifiers::kAltKey) !=
             0 &&
         event.IsOverLink();
}
```

Always returning  `event.IsOverLink();` allows selecting link text and, unlike userscript approach, it also prevents dragging of the already selected text, allowing to select again if you somehow messed on the first attempt for example. All of which results in the exact behavior I'm used to in old Opera :) <s>Though it still allows dragging the selection if that dragging was started on a plain text and not on a link, I could not find [yet?] a way to also fix this.</s> Text dragging is fixable by adding additional condition `!event.GetHitTestResult().IsSelected(event.GetHitTestLocation())` into `mouse_down_may_start_drag_` from `HandleMousePressEvent` in `third_party/blink/renderer/core/input/mouse_event_manager.cc`

Is this useful for ungoogled-chromium?

Unfortunately I couldn't hide it behind a flag. I used https://github.com/ungoogled-software/ungoogled-chromium/blob/master/patches/extra/ungoogled-chromium/add-flag-to-disable-sharing-hub.patch as an inspiration, and modified `selection_controller.cc` accordingly (pardon bad coding style, which is the result of experimenting):
```c
  if (base::CommandLine::ForCurrentProcess()->HasSwitch("disable-link-drag"))
  {return event.IsOverLink();}
  else {
   return (event.Event().GetModifiers() & WebInputEvent::Modifiers::kAltKey) !=
             0 &&
         event.IsOverLink();
   }
```
But `base::CommandLine::ForCurrentProcess()->HasSwitch("disable-link-drag")` seems to be always false regardless of the state of a flag. No idea what I'm doing wrong. :(

**Reply by PF4Public** [reply]
@Ahrotahn Would you be able to help me with this?

**Reply by Ahrotahn** [reply]
That section is in code that is spawned off of the main thread so it doesn't have access to commandline switches. You would have to implement this change as a feature flag instead.
